File: BackUpForABook/UIControls/Graph_framework/Graph/Graph/ChordGraph/CDConnection.m
//
//  CDConnection.m
//  ChordGraph
//
//  Created by Hui Wang on 6/27/14.
//  Copyright (c) 2014 AnnieLyticx. All rights reserved.
//

#import "CDConnection.h"
#import "CDCell.h"

@implementation CDConnection

-(id)initWith:(NSString*)lable
       ring_a:(int)ring_a
      slice_a:(int)slice_a
       cell_a:(int)cell_a
  intensity_a:(float)intensity_a
       ring_b:(int)ring_b
      slice_b:(int)slice_b
       cell_b:(int)cell_b
  intensity_b:(float)intensity_b
{
    self = [super init];
    
    _a.ring = ring_a;
    _a.slice = slice_a;
    _a.cell = cell_a;
    _a.intensity = intensity_a;

    _b.ring = ring_b;
    _b.slice = slice_b;
    _b.cell = cell_b;
    _b.intensity = intensity_b;
    
    
    _attrSmallFont = [NSMutableDictionary dictionary];
    _attrLargeFont = [NSMutableDictionary dictionary];
    
    [self setSmallFont:CFSTR("Helvetica") size:10 color:[UIColor darkTextColor]];
    [self setLargeFont:CFSTR("Helvetica") size:12 color:[UIColor darkTextColor]];
    
    _label = [[NSMutableString alloc] initWithString:lable];
    _aPath = [[UIBezierPath alloc] init];
    _highlight = FALSE;
    _showTip = TRUE;
    _largeFont = FALSE;
    
    return self;
}

// accessing node
-(CnnNode*)getNode:(int)ring
             slice:(int)slice
              cell:(int)cell
{
    if (_a.ring == ring &&
        _a.slice == slice &&
        _a.cell == cell)
    {
        return &(_a);
    }else if (_b.ring == ring &&
              _b.slice == slice &&
              _b.cell == cell)
    {
        return &(_b);
    }else
    {
        return nil;
    }
}

// check if is the connector between two given nodes
-(BOOL)isConnector:(CnnNode)a
                 b:(CnnNode)b
{
    if ([self getNode:a.ring slice:a.slice cell:a.cell] == nil)
    {
        return FALSE;
    }
    
    if ([self getNode:b.ring slice:b.slice cell:b.cell] == nil)
    {
        return FALSE;
    }
    return TRUE;
}

// call back from CDCell
-(BOOL)setNodeGuiSize:(id)cell
                 left:(float)left
                right:(float)right
{
    if ([cell isKindOfClass:[CDCell class]] != TRUE)
    {
        return FALSE;
    }
    
    if ([cell isEqual:_node_a] == TRUE)
    {
        _a.left = left;
        _a.right = right;
    }else if ([cell isEqual:_node_b] == TRUE)
    {
        _b.left = left;
        _b.right = right;
    }
        
    return TRUE;
}

// get intensity of the cell node
-(float)getIntensity:(id)cell
{
    if ([cell isKindOfClass:[CDCell class]] != TRUE)
    {
        return .0;
    }
    
    if ([cell isEqual:_node_a] == TRUE)
    {
        return _a.intensity;
    }else if ([cell isEqual:_node_b] == TRUE)
    {
        return _b.intensity;
    }
    
    return .0;
}

// set very basic attributes.  other cases please operate on the objects attrSmallFont and attrLargeFont directly
-(void)setSmallFont:(CFStringRef)font
               size:(float)size
              color:(UIColor*)color
{
    if (_attrSmallFont == nil)
    {
        _attrSmallFont = [NSMutableDictionary dictionary];
    }
    
    CTFontRef lbFont = CTFontCreateWithName(font, size, NULL);
    _attrSmallFont[NSFontAttributeName] = (__bridge id)lbFont;
    CFRelease(lbFont);
    _attrSmallFont[NSForegroundColorAttributeName] = (id)[color CGColor];
}

-(void)setLargeFont:(CFStringRef)font
               size:(float)size
              color:(UIColor*)color
{
    if (_attrLargeFont == nil)
    {
        _attrLargeFont = [NSMutableDictionary dictionary];
    }
    
    CTFontRef lbFont = CTFontCreateWithName(font, size, NULL);
    _attrLargeFont[NSFontAttributeName] = (__bridge id)lbFont;
    CFRelease(lbFont);
    _attrLargeFont[NSForegroundColorAttributeName] = (id)[color CGColor];
}

// reset gui view related properties
-(void)invalidate
{
    _a.painted = FALSE;
    _b.painted = FALSE;
}

// hit test
-(BOOL)hitTest:(CGPoint)pt
{
    return [_aPath containsPoint:pt];
}

@end

File: BackUpForABook/UIControls/CDClasses_swift/CDConnection.swift
//
//  CDConnection.swift
//  ChordTest
//
//  Created by dingf on 16/12/12.
//  Copyright © 2016年 MH.dingf. All rights reserved.
//

import UIKit
struct CnnNode {
    // ring index
    var ring :Int!
    // slice index
    var slice :Int!
    // cell index
    var cell :Int!
    // connectivity intensity in terms of percentage
    var intensity :CGFloat!
    // start / end angle degrees
    var left :CGFloat = 0
    var right :CGFloat = 0
    // true if already painted
    var painted :Bool!
}

// connection to other element (ring index, slice index, cell index)
// used for connection between two cell connectors
// each cell may have array of indexs to CDConnections that "connects" two cells

class CDConnection: NSObject {
    // node a and b of connector
    var cnnNodeA = CnnNode()
    var node_a :AnyObject!
    var cnnNodeB = CnnNode()
    var node_b :AnyObject!
    // small font for showing tips for groups of highlighted connectors
    var attrSmallFont :NSMutableDictionary!
    // large font
    var attrLargeFont :NSMutableDictionary!
    // label or connector text details
    var label :String!
    // information for visual presentation
    var aPath :UIBezierPath!
    // if show as highlighted
    var highlight :Bool!
    // if show tip
    var showTip :Bool!
    // if display in large font
    var largeFont :Bool!
    
    init(cnnLabel: String,
           ring_a: Int,
          slice_a: Int,
           cell_a: Int,
      intensity_a: CGFloat,
           ring_b: Int,
          slice_b: Int,
           cell_b: Int,
      intensity_b: CGFloat)
    {
        super.init()
        cnnNodeA.ring = ring_a
        cnnNodeA.slice = slice_a
        cnnNodeA.cell = cell_a
        cnnNodeA.intensity = intensity_a
        
        cnnNodeB.ring = ring_b
        cnnNodeB.slice = slice_b
        cnnNodeB.cell = cell_b
        cnnNodeB.intensity = intensity_b
        
        attrSmallFont = NSMutableDictionary()
        attrLargeFont = NSMutableDictionary()
        // set attrSmallFont and attr
        setSmall("Helverica" as CFString, size: 10, color:UIColor.darkText)
        setLarge("Helverica" as CFString, size: 12, color: UIColor.darkText)
        
        label = cnnLabel
        aPath = UIBezierPath()
        highlight = false
        showTip = true
        largeFont = false
    }
    
    // set very basic attributes. other cases please operate on the objects attrSmallFont and attrLargeFont directly
    func setSmall(_ font:CFString,size:CGFloat,color:UIColor){
        if attrSmallFont == nil{
            attrSmallFont = NSMutableDictionary()
        }
        let lbFont = CTFontCreateWithName(font, size, nil)
        attrSmallFont[NSFontAttributeName] = lbFont
        attrSmallFont[NSForegroundColorAttributeName] = color.cgColor
    }
    
    func setLarge(_ font:CFString,size:CGFloat,color:UIColor){
        if attrLargeFont == nil{
            attrLargeFont = NSMutableDictionary()
        }
        let lbFont = CTFontCreateWithName(font, size, nil)
        attrLargeFont[NSFontAttributeName] = lbFont
        attrLargeFont[NSForegroundColorAttributeName] = color.cgColor
    }
    
    // get intensity of the cell node
    func getIntensity(cell:AnyObject) -> CGFloat{
        if cell.isKind(of: CDCell.self) != true{
            return 0.0
        }
        if cell.isEqual(node_a) == true{
            return cnnNodeA.intensity
        }else if cell.isEqual(node_b) == true{
            return cnnNodeB.intensity
        }
        return 0.0
    }
    
    // set the left and right of connector's point A and point B ,it is according to the parameter "cell"
    func setNodeGuiSize(cell:AnyObject,left:CGFloat,right:CGFloat) -> Bool{
        if cell.isKind(of: CDCell.self) != true{
            return false
        }
        if cell.isEqual(node_a) == true{
            cnnNodeA.left = left
            cnnNodeA.right = right
        }else if cell.isEqual(node_b) == true{
            cnnNodeB.left = left
            cnnNodeB.right = right
        }
        return true
    }
    
    // hit test 
    func hitTest(_ pt:CGPoint) -> Bool{
        return aPath.contains(pt)
    }
}
